### PR TITLE
Renderer basic

### PR DESCRIPTION
**Pull request by Saiel**
First implementation that had been done for presentation

### CODE REVIEW

File: src/renderer/src/renderer/vertexarray.h [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+	virtual void Bind() const = 0;
+	virtual void Unbind() const = 0;
+
+	virtual void SetIndexBuffer(std::unique_ptr<IndexBuffer>&& ib) = 0;
+	virtual const IndexBuffer& indexBuffer() const = 0;
+
+	static std::unique_ptr<VertexArray> Create();
+
+	static std::unique_ptr<VertexArray> Create(std::unique_ptr<VertexBuffer>&& vb,
+	                                           std::unique_ptr<IndexBuffer>&& ib);
```
Comment (CapsRePeaT) [inline src/renderer/src/renderer/vertexarray.h:20]: vs

File: src/renderer/src/renderer/vertexbuffer.cpp [hunk 2]
```diff
@@ -0,0 +1,30 @@
+#include "vertexbuffer.h"
+
+#include "platform/opengl/vertexbuffer.h"
+#include "renderer_api.h"
+
+std::unique_ptr<VertexBuffer> VertexBuffer::Create(size_t size) {
```
Comment (CapsRePeaT) [inline src/renderer/src/renderer/vertexbuffer.cpp:6]: const

File: src/renderer/src/renderer/vertexbuffer.cpp [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+			return nullptr;
+		case RendererAPI::API::OpenGL:
+			return std::make_unique<OpenGLVertexBuffer>(size);
+	}
+	assert(false);
+	return nullptr;
+}
+
+std::unique_ptr<VertexBuffer> VertexBuffer::Create(
+		const void* data, size_t size,
```
Comment (CapsRePeaT) [inline src/renderer/src/renderer/vertexbuffer.cpp:19]: const

File: src/renderer/src/renderer/vertexbuffer.cpp [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+		std::unique_ptr<VertexBufferLayout>&& layout) {
+	switch (RendererAPI::instance()->api()) {
+		case RendererAPI::API::None:
+			assert(false && "Renderer::API::None is not currently supported");
+			return nullptr;
+		case RendererAPI::API::OpenGL:
+			return std::make_unique<OpenGLVertexBuffer>(data, size, std::move(layout));
+	}
+	assert(false);
+	return nullptr;
```
Comment (CapsRePeaT) [inline src/renderer/src/renderer/vertexbuffer.cpp:29]: this part almost completelly duplicates 4 times, 2 times in this file and 2 times in vertex_array.cpp please unify it

File: src/renderer/src/renderer/vertexbuffer.h [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+	template <typename T, size_t size>
+	inline void SetData(const std::array<T, size>& data,
+			std::unique_ptr<VertexBufferLayout>&& layout) {
+		SetData(size * sizeof(T), data.data(), std::move(layout));
+	}
+
+	// Will be deleted and statically resolved via reflection and vertex structs
+	virtual void SetLayout(std::unique_ptr<VertexBufferLayout>&& layout) = 0;
+	[[nodiscard]] virtual const VertexBufferLayout& layout() const = 0;
+
```
Comment (CapsRePeaT) [inline src/renderer/src/renderer/vertexbuffer.h:25]: vs

File: src/renderer/src/renderer/vertexbufferlayout.cpp [hunk 6]
```diff
@@ -0,0 +1,11 @@
+#include "vertexbufferlayout.h"
+
+unsigned int VertexBufferLayout::VertexBufferElement::size() const {
+	switch (type) {
+		case GL_FLOAT:          return sizeof(float);
+		case GL_UNSIGNED_INT:   return sizeof(uint32_t);
+		case GL_BYTE:           return sizeof(int8_t);
+	}
+
```
Comment (CapsRePeaT) [inline src/renderer/src/renderer/vertexbufferlayout.cpp:9]: vs

File: src/renderer/src/renderer/vertexbufferlayout.h [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+	template <typename T>
+	void Push(unsigned int count) {
+		elements_.push_back({count, TypeToGlEnum<T>(), sizeof(T) == 1});
+		stride_ += sizeof(T) * count;
+	}
+	[[nodiscard]] inline int32_t stride() const { return stride_; }
+
+	[[nodiscard]] inline const VertexBufferElements& elements() const {
+		return elements_;
+	}
```
Comment (CapsRePeaT) [inline src/renderer/src/renderer/vertexbufferlayout.h:39]: vs

File: src/ui/src/metadata_widget.cpp [hunk 8]
```diff
@@ -0,0 +1,14 @@
+#include "metadata_widget.h"
+#include "ui_metadata_widget.h"
+
+MetadataWidget::MetadataWidget(QWidget *parent) :
+  QDockWidget(parent),
+  ui(new Ui::MetadataWidget)
+{
```
Comment (CapsRePeaT) [inline src/ui/src/metadata_widget.cpp:7]: ui(new Ui::MetadataWidget) {

File: src/ui/src/metadata_widget.cpp [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+  ui->setupUi(this);
+}
+
+MetadataWidget::~MetadataWidget()
+{
```
Comment (CapsRePeaT) [inline src/ui/src/metadata_widget.cpp:12]: MetadataWidget::~MetadataWidget() {

File: src/ui/src/metadata_widget.h [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+class MetadataWidget : public QDockWidget
+{
+  Q_OBJECT
+
+public:
+  explicit MetadataWidget(QWidget *parent = nullptr);
+  ~MetadataWidget();
+
+private:
+  Ui::MetadataWidget *ui;
```
Comment (CapsRePeaT) [inline src/ui/src/metadata_widget.h:19]: * ui_

File: src/ui/src/renderer_widget.cpp [hunk 11]
```diff
@@ -1,14 +1,69 @@
...
+QSurfaceFormat defaultFormat() {
+	auto surface_format = QSurfaceFormat();
+	surface_format.setVersion(4, 3);
+	surface_format.setProfile(QSurfaceFormat::CoreProfile);
+	surface_format.setRenderableType(QSurfaceFormat::OpenGL);
+	surface_format.setSwapBehavior(QSurfaceFormat::TripleBuffer);
+	surface_format.setSwapInterval(1);
+	surface_format.setDepthBufferSize(24);
+	surface_format.setStencilBufferSize(8);
+
```
Comment (CapsRePeaT) [inline src/ui/src/renderer_widget.cpp:54]: vs

File: src/ui/src/renderer_widget.cpp [hunk 12]
```diff
@@ -1,14 +1,69 @@
...
+// TODO: maybe need to delegate tweaks to renderer
+QSurfaceFormat defaultFormat() {
+	auto surface_format = QSurfaceFormat();
+	surface_format.setVersion(4, 3);
+	surface_format.setProfile(QSurfaceFormat::CoreProfile);
+	surface_format.setRenderableType(QSurfaceFormat::OpenGL);
+	surface_format.setSwapBehavior(QSurfaceFormat::TripleBuffer);
+	surface_format.setSwapInterval(1);
+	surface_format.setDepthBufferSize(24);
+	surface_format.setStencilBufferSize(8);
```
Comment (CapsRePeaT) [inline src/ui/src/renderer_widget.cpp:53]: better quick description of magic numbers

File: src/ui/src/renderer_widget.cpp [hunk 13]
```diff
@@ -1,14 +1,69 @@
...
+QSurfaceFormat debugFormat() {
+	auto surface_format = QSurfaceFormat(QSurfaceFormat::DebugContext);
+	surface_format.setVersion(4, 3);
+	surface_format.setProfile(QSurfaceFormat::CoreProfile);
+	surface_format.setRenderableType(QSurfaceFormat::OpenGL);
+	surface_format.setSwapBehavior(QSurfaceFormat::TripleBuffer);
+	surface_format.setSwapInterval(1);
+	surface_format.setDepthBufferSize(24);
+	surface_format.setStencilBufferSize(8);
+
```
Comment (CapsRePeaT) [inline src/ui/src/renderer_widget.cpp:67]: vs

File: src/ui/src/renderer_widget.cpp [hunk 14]
```diff
@@ -1,14 +1,69 @@
...
+QSurfaceFormat debugFormat() {
+	auto surface_format = QSurfaceFormat(QSurfaceFormat::DebugContext);
+	surface_format.setVersion(4, 3);
+	surface_format.setProfile(QSurfaceFormat::CoreProfile);
+	surface_format.setRenderableType(QSurfaceFormat::OpenGL);
+	surface_format.setSwapBehavior(QSurfaceFormat::TripleBuffer);
+	surface_format.setSwapInterval(1);
+	surface_format.setDepthBufferSize(24);
+	surface_format.setStencilBufferSize(8);
```
Comment (CapsRePeaT) [inline src/ui/src/renderer_widget.cpp:66]: code dume, just pass QSurfaceFormat as an arg

File: src/ui/src/renderer_widget.h [hunk 15]
```diff
@@ -1,35 +1,41 @@
...
+	explicit RendererWidget(QWidget* parent = nullptr);
+	~RendererWidget() override;
+
+	void RenderShapes(const Core::Shapes& shapes) {
+		assert(!shapes.empty() && "no shapes received");
+		makeCurrent();
+		renderer_->ClearScene();
+		renderer_->RenderShapes(shapes);
+		doneCurrent();
+	};
```
Comment (CapsRePeaT) [inline src/ui/src/renderer_widget.h:24]: move to cpp please, too many lines for header

File: src/ui/src/renderer_widget.h [hunk 16]
```diff
@@ -1,35 +1,41 @@
...
+	void UpdateVisualisationOptions(
+			const VisualisationOptions& visualisation_options);
+
```
Comment (CapsRePeaT) [inline src/ui/src/renderer_widget.h:27]: vs

File: src/ui/src/shape_list_widget.cpp [hunk 17]
```diff
@@ -0,0 +1,14 @@
+#include "shape_list_widget.h"
+#include "ui_shape_list_widget.h"
+
+ShapeListWidget::ShapeListWidget(QWidget *parent) :
+  QDockWidget(parent),
+  ui(new Ui::ShapeListWidget)
+{
```
Comment (CapsRePeaT) [inline src/ui/src/shape_list_widget.cpp:7]: ui(new Ui::ShapeListWidget) {

File: src/ui/src/shape_list_widget.cpp [hunk 18]
```diff
@@ -0,0 +1,14 @@
...
+  ui->setupUi(this);
+}
+
+ShapeListWidget::~ShapeListWidget()
```
Comment (CapsRePeaT) [inline src/ui/src/shape_list_widget.cpp:11]: ShapeListWidget::~ShapeListWidget() {

File: src/ui/src/shape_list_widget.h [hunk 19]
```diff
@@ -0,0 +1,22 @@
...
+class ShapeListWidget : public QDockWidget
+{
+  Q_OBJECT
+
+public:
+  explicit ShapeListWidget(QWidget *parent = nullptr);
+  ~ShapeListWidget();
+
+private:
+  Ui::ShapeListWidget *ui;
```
Comment (CapsRePeaT) [inline src/ui/src/shape_list_widget.h:19]: Ui::ShapeListWidget* ui_

File: src/ui/src/shape_list_widget.h [hunk 20]
```diff
@@ -0,0 +1,22 @@
...
+namespace Ui {
+class ShapeListWidget;
+}
+
+class ShapeListWidget : public QDockWidget
+{
+  Q_OBJECT
+
+public:
+  explicit ShapeListWidget(QWidget *parent = nullptr);
```
Comment (CapsRePeaT) [inline src/ui/src/shape_list_widget.h:15]: QWidget* parent

File: src/ui/src/shape_list_widget.cpp [hunk 21]
```diff
@@ -0,0 +1,14 @@
+#include "shape_list_widget.h"
+#include "ui_shape_list_widget.h"
+
+ShapeListWidget::ShapeListWidget(QWidget *parent) :
+  QDockWidget(parent),
+  ui(new Ui::ShapeListWidget)
```
Comment (CapsRePeaT) [inline src/ui/src/shape_list_widget.cpp:6]: in one line plz

File: src/ui/src/main.cpp [hunk 22]
```diff
@@ -1,12 +1,14 @@
+#include <spdlog/spdlog.h>
+
 #include <QApplication>
 #include <QLabel>
 
 #include "mainwindow.h"
 
-int main(int argc, char** argv)
-{
+int main(int argc, char** argv) {
```
Comment (CapsRePeaT) [inline src/ui/src/main.cpp:8]: just reminder that Alexey QT

**Review by CapsRePeaT — CHANGES_REQUESTED** [review]

File: src/renderer/include/i_renderer.h [hunk 23]
```diff
@@ -1,12 +1,47 @@
 #pragma once
 
 #include "common.h"
+#include "core.h"
 
+// TODO: rename to smth as Scene
```
[the remark this answers is not in the document]
Comment (Saiel) [reply src/renderer/include/i_renderer.h:6]: `Scene`, `SceneRenderer`, `SceneViewport` or , idk. Pure renderer is part of this module (luck that their names won't collide) and this class is not only about rendering but also to interact with objects (e.g. to select them) or scene itself (add/remove objects)

File: src/renderer/res/shaders/heatmap.fs [hunk 24]
```diff
@@ -0,0 +1,28 @@
+#version 430
+
+in vec2 frag_UVCoordinates;
+out vec4 out_Color;
+
+uniform vec2 u_TemperatureRange;
+uniform mat2x3 u_ColorRange;
+
+float sqr(float x) { return x*x; }
```
[the remark this answers is not in the document]
Comment (Saiel) [reply src/renderer/res/shaders/heatmap.fs:9]: Spaces around arithmetic operators are not mandatory

File: src/renderer/src/gl_renderer.cpp [hunk 25]
```diff
@@ -0,0 +1,89 @@
...
+#include "renderer/renderer_api.h"
+
+struct GLRenderer::GLRendererData {
+	std::unique_ptr<HeatmapMaterial> heatmap_material;
+	std::vector<std::shared_ptr<SceneShape>> scene_shapes;
+	std::unique_ptr<OrthographicCamera> camera;
+};
+
+GLRenderer::GLRenderer() : data_(std::make_unique<GLRendererData>()) {}
+
```
[the remark this answers is not in the document]
Comment (Saiel) [reply src/renderer/src/gl_renderer.cpp:23]: function definition

File: src/renderer/src/gl_renderer.cpp [hunk 26]
```diff
@@ -0,0 +1,89 @@
...
+GLRenderer::~GLRenderer() { ClearResourcesImpl(); }
+
+void GLRenderer::Initialize(int w, int h) {
+	// OpenGL initialization
+	if (!gladLoadGL()) spdlog::error("Failed to initialize opengl functions");
+	spdlog::debug("GLAD initialized");
```
Comment (Saiel) [inline src/renderer/src/gl_renderer.cpp:29]: Condition is trivial and no line breaks is permitted by Google style

File: src/renderer/src/platform/opengl/vertexbuffer.h [hunk 27]
```diff
@@ -0,0 +1,38 @@
...
+		layout_ = std::move(layout);
+	}
+	[[nodiscard]] const VertexBufferLayout& layout() const { return *layout_; }
+
+	void Bind() const final;
+	void Unbind() const final;
+
+	void SetData(const void* data, size_t size,
+	             std::unique_ptr<VertexBufferLayout>&& layout) final;
+
```
[the remark this answers is not in the document]
Comment (Saiel) [reply src/renderer/src/platform/opengl/vertexbuffer.h:32]: vs before `private:`

File: src/renderer/src/application/heatmap_material.h [hunk 28]
```diff
@@ -0,0 +1,42 @@
+#pragma once
+
+#include <glm/glm.hpp>
+#include <memory>
+
```
[the remark this answers is not in the document]
Comment (Saiel) [reply src/renderer/src/application/heatmap_material.h:5]: Discussed, that spaces here are acceptable

File: src/renderer/src/application/heatmap_material.h [hunk 29]
```diff
@@ -0,0 +1,42 @@
...
+		temperature_range_[1] = max;
+	}
+	void SetColorRange(const glm::vec3& min, const glm::vec3& max) {
+		color_range_[0] = min;
+		color_range_[1] = max;
+	}
+	void SetShader(const std::shared_ptr<ShaderProgram>& shader) {
+		shader_ = shader;
+	}
+
```
[the remark this answers is not in the document]
Comment (Saiel) [reply src/renderer/src/application/heatmap_material.h:23]: no, 7 extra symbols

(changed to take by-value and `std::move` to `shader_` for explicity fo client code, that there will be ref count incrementing and to allow for std::move for avoiding that)

File: src/renderer/src/application/scene_shape.cpp [hunk 30]
```diff
@@ -0,0 +1,69 @@
+#include "scene_shape.h"
+
+#include <memory>
+
```
[the remark this answers is not in the document]
Comment (Saiel) [reply src/renderer/src/application/scene_shape.cpp:4]: see heatmap_material.h answer

File: src/renderer/src/platform/opengl/indexbuffer.cpp [hunk 31]
```diff
@@ -0,0 +1,41 @@
...
+OpenGLIndexBuffer::~OpenGLIndexBuffer() { glDeleteBuffers(1, &id_); }
+
+OpenGLIndexBuffer::OpenGLIndexBuffer(OpenGLIndexBuffer&& other) noexcept {
+	if (this == &other) return;
+	*this = std::move(other);
+}
+
+OpenGLIndexBuffer& OpenGLIndexBuffer::operator=(
+		OpenGLIndexBuffer&& other) noexcept {
+	if (this == &other) return *this;
```
[the remark this answers is not in the document]
Comment (Saiel) [reply src/renderer/src/platform/opengl/indexbuffer.cpp:22]: no line break is appropriate by google style as statement is brief and short enough https://google.github.io/styleguide/cppguide.html#Conditionals

Indents (both vertical and horizontal) are the main guides for code structure. IMO if you see `if` and next line is not indented, it's obvious the line isn't a part of it and you should see at the closing bracket of condition to find a statement

File: src/renderer/src/platform/opengl/indexbuffer.cpp [hunk 32]
```diff
@@ -0,0 +1,41 @@
...
+	// Delete owned buffer
+	glDeleteBuffers(1, &id_);
+	// Assign moving buffer data
+	id_ = other.id_;
+	count_ = other.count_;
+	// Invalidate moving buffer
+	other.id_ = 0;
```
[the remark this answers is not in the document]
Comment (Saiel) [reply src/renderer/src/platform/opengl/indexbuffer.cpp:29]: in context of OpenGL (and others renderer APIs) it's well defined, that id of existing renderer object is >=1 and 0 is value for safe invalidation

File: src/renderer/src/platform/opengl/renderer_api.cpp [hunk 33]
```diff
@@ -0,0 +1,60 @@
+#include "renderer_api.h"
+
+#include <glad/glad.h>
+#include <spdlog/spdlog.h>
+
+void OpenGLMessageCallback(unsigned source, unsigned type, unsigned id,
```
[the remark this answers is not in the document]
Comment (Saiel) [reply src/renderer/src/platform/opengl/renderer_api.cpp:6]: Changed signature to one defined in OpenGL documentation for portability

File: src/renderer/src/platform/opengl/vertexarray.cpp [hunk 34]
```diff
@@ -0,0 +1,65 @@
+#include "vertexarray.h"
+
+#include <glad/glad.h>
+
```
[the remark this answers is not in the document]
Comment (Saiel) [reply src/renderer/src/platform/opengl/vertexarray.cpp:4]: see `heatmap_material.h`

File: src/renderer/src/platform/opengl/vertexarray.cpp [hunk 35]
```diff
@@ -0,0 +1,65 @@
...
+                                     std::unique_ptr<IndexBuffer>&& ib) {
+	glGenVertexArrays(1, &id_);
+	SetBuffer(std::move(vb));
+	SetIndexBuffer(std::move(ib));
+}
+
+OpenGLVertexArray::~OpenGLVertexArray() { glDeleteVertexArrays(1, &id_); }
+
+OpenGLVertexArray::OpenGLVertexArray(OpenGLVertexArray&& other) noexcept {
+	if (this == &other) return;
```
[the remark this answers is not in the document]
Comment (Saiel) [reply src/renderer/src/platform/opengl/vertexarray.cpp:17]: see opengl/indexbuffer.cpp

File: src/renderer/src/platform/opengl/vertexarray.cpp [hunk 36]
```diff
@@ -0,0 +1,65 @@
...
+	*this = std::move(other);
+}
+
+OpenGLVertexArray& OpenGLVertexArray::operator=(
+		OpenGLVertexArray&& other) noexcept {
+	if (this == &other) return *this;
```
[the remark this answers is not in the document]
Comment (Saiel) [reply src/renderer/src/platform/opengl/vertexarray.cpp:23]: see opengl/indexbuffer.cpp

File: src/renderer/src/platform/opengl/vertexarray.cpp [hunk 37]
```diff
@@ -0,0 +1,65 @@
...
+	// Delete owned array
+	glDeleteVertexArrays(1, &id_);
+	// Assign moving array data
+	id_ = other.id_;
+	SetBuffer(std::move(other.vbo_));
+	SetIndexBuffer(std::move(other.ib_));
+	// Invalidate moving array
+	other.id_ = 0;
```
[the remark this answers is not in the document]
Comment (Saiel) [reply src/renderer/src/platform/opengl/vertexarray.cpp:31]: see opengl/indexbuffer.cpp

File: src/renderer/src/platform/opengl/vertexarray.cpp [hunk 38]
```diff
@@ -0,0 +1,65 @@
...
+void OpenGLVertexArray::SetBuffer(std::unique_ptr<VertexBuffer>&& vb) {
+	Bind();
+	vb->Bind();
+	auto& layout = vb->layout();
+
+	char* offset = nullptr;  // char* is compromise between conversion to void*
+	                         // and pointer arithmetic
+	const auto& elements = layout.elements();
+
+	for (unsigned int i = 0; i < elements.size(); ++i) {
```
Comment (Saiel) [inline src/renderer/src/platform/opengl/vertexarray.cpp:45]: I want to leave this vs for better clearance of loop definition

File: src/renderer/src/platform/opengl/vertexarray.cpp [hunk 39]
```diff
@@ -0,0 +1,65 @@
...
+		const auto& element = elements[i];
+		glEnableVertexAttribArray(i);
+		glVertexAttribPointer(i, element.count, element.type, element.normalized,
```
Comment (Saiel) [inline src/renderer/src/platform/opengl/vertexarray.cpp:48]: such a cool semantic

File: src/renderer/src/platform/opengl/vertexbuffer.cpp [hunk 40]
```diff
@@ -0,0 +1,55 @@
+#include "vertexbuffer.h"
+
+#include <glad/glad.h>
+
+#include <memory>
+
+#include "renderer/vertexbufferlayout.h"
```
[the remark this answers is not in the document]
Comment (Saiel) [reply src/renderer/src/platform/opengl/vertexbuffer.cpp:7]: see `heatmap_material.h`

File: src/renderer/src/platform/opengl/vertexbuffer.cpp [hunk 41]
```diff
@@ -0,0 +1,55 @@
...
+	glGenBuffers(1, &id_);
+	Bind();
+	glBufferData(GL_ARRAY_BUFFER, size, data, GL_STATIC_DRAW);
+	layout_ = std::move(layout);
+}
+
+OpenGLVertexBuffer::~OpenGLVertexBuffer() { glDeleteBuffers(1, &id_); }
+
+OpenGLVertexBuffer::OpenGLVertexBuffer(OpenGLVertexBuffer&& other) noexcept {
+	if (this == &other) return;
```
[the remark this answers is not in the document]
Comment (Saiel) [reply src/renderer/src/platform/opengl/vertexbuffer.cpp:27]: see `indexbuffer.cpp`

File: src/renderer/src/platform/opengl/vertexbuffer.cpp [hunk 42]
```diff
@@ -0,0 +1,55 @@
...
+	*this = std::move(other);
+}
+
+OpenGLVertexBuffer& OpenGLVertexBuffer::operator=(
+		OpenGLVertexBuffer&& other) noexcept {
+	if (this == &other) return *this;
```
[the remark this answers is not in the document]
Comment (Saiel) [reply src/renderer/src/platform/opengl/vertexbuffer.cpp:33]: see `indexbuffer.cpp`

File: src/renderer/src/platform/opengl/vertexbuffer.cpp [hunk 43]
```diff
@@ -0,0 +1,55 @@
...
+	// Delete owned buffer
+	glDeleteBuffers(1, &id_);
+	// Assign moving buffer data
+	id_ = other.id_;
+	layout_ = std::move(other.layout_);
+	// Invalidate moving buffer
+	other.id_ = 0;
```
[the remark this answers is not in the document]
Comment (Saiel) [reply src/renderer/src/platform/opengl/vertexbuffer.cpp:40]: see `indexbuffer.cpp`

File: src/ui/src/mainwindow.cpp [hunk 44]
```diff
@@ -1,39 +1,44 @@
...
+	visualization_options_->setWindowTitle("Options");
+	// signals and slots connection
+	connect(ui_->load_file_btn, &QAction::triggered, this,
+	        &MainWindow::OnLoadFileBtnPressed);
 }
 
-MainWindow::~MainWindow() {
-  delete ui_;
-}
+MainWindow::~MainWindow() { delete ui_; }
```
Comment (CapsRePeaT) [inline src/ui/src/mainwindow.cpp:32]: move to header plz

File: src/ui/src/metadata_widget.h [hunk 45]
```diff
@@ -0,0 +1,18 @@
...
+
+class MetadataWidget : public QDockWidget {
+	Q_OBJECT
+
+ public:
+	explicit MetadataWidget(QWidget* parent = nullptr);
+	~MetadataWidget();
+
+ private:
+	Ui::MetadataWidget* ui;
```
Comment (CapsRePeaT) [inline src/ui/src/metadata_widget.h:17]: ui_

File: src/ui/src/renderer_widget.cpp [hunk 46]
```diff
@@ -1,14 +1,89 @@
...
 
-RendererWidget::~RendererWidget()
-{
-    delete ui_;
+void RendererWidget::paintGL() {
+	renderer_->RenderFrame();
+	update();
+}
+
+void RendererWidget::resizeGL(int w, int h) {
```
Comment (CapsRePeaT) [inline src/ui/src/renderer_widget.cpp:59]: const

File: src/ui/src/shape_list_widget.cpp [hunk 47]
```diff
@@ -0,0 +1,10 @@
+#include "shape_list_widget.h"
+
+#include "ui_shape_list_widget.h"
+
+ShapeListWidget::ShapeListWidget(QWidget* parent)
+		: QDockWidget(parent), ui(new Ui::ShapeListWidget) {
+	ui->setupUi(this);
```
Comment (CapsRePeaT) [inline src/ui/src/shape_list_widget.cpp:7]: ui_

**Review by CapsRePeaT — APPROVED** [review]

File: src/renderer/src/renderer/vertexarray.cpp [hunk 48]
```diff
@@ -0,0 +1,31 @@
...
+	}
+	assert(false);
+	return nullptr;
+}
+
+std::unique_ptr<VertexArray> VertexArray::Create(
+		std::unique_ptr<VertexBuffer>&& vb, std::unique_ptr<IndexBuffer>&& ib) {
+	switch (RendererAPI::instance()->api()) {
+		case RendererAPI::API::None:
+			assert(false && "Renderer::API::None is not currently supported");
```
[the remark this answers is not in the document]
Comment (Saiel) [reply src/renderer/src/renderer/vertexarray.cpp:25]: All usages will be separated in one abstract factory. TODO

File: src/renderer/include/i_renderer.h [hunk 49]
```diff
@@ -1,12 +1,47 @@
...
 class IRenderer {
```
[the remark this answers is not in the document]
Comment (Saiel) [reply src/renderer/include/i_renderer.h:7]: TODOed

File: src/ui/src/mainwindow.cpp (hunk 44 above)
Comment (Saiel) [reply src/ui/src/mainwindow.cpp:32]: ui_ definition places in ui_mainwindow.h, that we won't move to header. So type UI::MainWindow is incomplete in header and we can't delete instatnces of it as it cause UB

File: src/renderer/src/platform/opengl/indexbuffer.cpp [hunk 50]
```diff
@@ -0,0 +1,41 @@
...
+
+OpenGLIndexBuffer::OpenGLIndexBuffer(const uint32_t* data, const size_t count)
+		: count_(count) {
+	glGenBuffers(consts::buffer_count_one, &id_);
+	Bind();
+	glBufferData(GL_ELEMENT_ARRAY_BUFFER, count * sizeof(uint32_t), data,
+	             GL_STATIC_DRAW);
+}
+
+OpenGLIndexBuffer::~OpenGLIndexBuffer() { glDeleteBuffers(consts::buffer_count_one, &id_); }
```
[the remark this answers is not in the document]
Comment (Saiel) [reply src/renderer/src/platform/opengl/indexbuffer.cpp:15]: I don't want to include glad in headers, so no

File: src/renderer/include/i_renderer.h (hunk 23 above)
[the remark this answers is not in the document]
Comment (Saiel) [reply src/renderer/include/i_renderer.h:6]: TODOed

File: src/renderer/include/i_renderer.h (hunk 49 above)
[the remark this answers is not in the document]
Comment (Saiel) [reply src/renderer/include/i_renderer.h:7]: TODOed

File: src/common/include/common.h [hunk 51]
```diff
@@ -1,33 +1,43 @@
...
-template<int dim>
+// represents 3d box //should we use dimension as template parameter?
+// array of min and max cords by dimension, by default values from numeric
+// limits
+template <int dim>
 class Box {
  public:
-  bool undefined() const { return coordinates_ == DefaultCoordinates(); }
-  Box() : coordinates_(DefaultCoordinates()) {}
+	using Values = std::array<std::pair<float, float>, dim>;
```
[the remark this answers is not in the document]
Comment (Saiel) [reply src/common/include/common.h:16]: It is returned from coordinates on 24 LOC. I don't see any problem with it

File: src/renderer/src/renderer/vertexbuffer.cpp (hunk 4 above)
Comment (Saiel) [reply src/renderer/src/renderer/vertexbuffer.cpp]: Will be in one abstract factory, not in this pr

File: src/common/include/common.h (hunk 51 above)
[the remark this answers is not in the document]
Comment (Saiel) [reply src/common/include/common.h:16]: TODOed

**Pull request merged**